Scannergates 2.0

Axanthic opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Scanner Gates can now have assemblies attached to them! Allow you to trigger anything you want with them!

![XGWvtJbPH9](https://user-images.githubusercontent.com/19522166/88450160-e4b8a880-ce1a-11ea-83c6-dac1cb35cd06.gif)


There are 5 wires on them now.
Successful Scan
Failed Scan
Toggle Signaling
Disable Signaling
Enabling Signaling

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->


## Why It's Good For The Game
Imagine the possibilities! Traps! Security Checkpoints! 
 
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Scanner Gates can now trigger assemblies!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

SheldonCurtiss committed:
On branch improvedscanners
 Changes to be committed:
	modified:   beestation.dme
	modified:   code/__DEFINES/wires.dm
	new file:   code/datums/wires/scan_gate.dm
	modified:   code/game/machinery/scan_gate.dm

ike709 commented on code/game/machinery/scan_gate.dm:
```diff
@@ -164,6 +180,15 @@
+		if(!ignore_signals)
+			color = wires.get_color_of_wire(WIRE_FAIL)
+			var/obj/item/assembly/S = wires.get_attached(color)
+			S.activate()
```
```suggestion
			if(istype(S))
			    S.activate()
```

ike709 commented on code/game/machinery/scan_gate.dm:
```diff
@@ -164,6 +180,15 @@
+		if(!ignore_signals)
+			color = wires.get_color_of_wire(WIRE_PASS)
+			var/obj/item/assembly/S = wires.get_attached(color)
+			S.activate()
```
```suggestion
          if(istype(S))
			    S.activate()

```

SheldonCurtiss committed:
Update code/game/machinery/scan_gate.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

SheldonCurtiss committed:
Update code/game/machinery/scan_gate.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

ghost commented:
@ike709 I did as you ask. 👀

francinum reviewed (dismissed):
My reviews don't even count

ghost commented:
@francinum They count to me

PowerfulBacon commented:
Wow, nice.
Make sure you update the hacking page on our wiki.

ike709 commented:
@francinum @PowerfulBacon neither of you pointed out that this doesn't even compile

PowerfulBacon commented:
Oh they used spaces instead of tabs?
I was on mobile when I looked at this and didnt scroll down.
Anyway, it doesn't compile  @axanthic

ghost commented:
> @francinum @PowerfulBacon neither of you pointed out that this doesn't even compile

My bad guys - Thing I messed up on the Push will fix in a bit

SheldonCurtiss committed:
Update scan_gate.dm

ghost commented:
> @francinum @PowerfulBacon neither of you pointed out that this doesn't even compile

Ike broke it with his cool changes >:C I take 0 responsibility for this. Anyhow should be fixed and set to go.

ghost commented:
> Wow, nice.
> Make sure you update the hacking page on our wiki.

Is this something I have to do?
I'm unsure what changes should even be made since as far as I'm aware the wire setup is unique to this device. So I don't know if having a specific section just for it makes sense?

SheldonCurtiss committed:
Update scan_gate.dm

SheldonCurtiss committed:
Update scan_gate.dm

SheldonCurtiss committed:
Update scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

ghost commented:
I'm a dumbass - Should be set to go.

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/datums/wires/scan_gate.dm
	modified:   code/game/machinery/scan_gate.dm

ghost commented:
Updated the names of the variables and added a general trigger wire.

ike709 commented:
What if the wires are cut?

qwertyquerty commented on code/game/machinery/scan_gate.dm:
```diff
@@ -160,11 +177,26 @@
 
+	if(!ignore_signals)
+		color = wires.get_color_of_wire(WIRE_PROCEED)
+		var/obj/item/assembly/S = wires.get_attached(color)
```
You do these 3 lines 3 times in 3 different places. Just do them once at the end.

qwertyquerty commented on code/game/machinery/scan_gate.dm:
```diff
@@ -50,12 +55,18 @@
 		. += "<span class='notice'>The control panel is unlocked. Swipe an ID to lock it.</span>"
 
+/obj/machinery/scanner_gate/update_icon_state()
+	if(panel_open)
```
What is the point of this

ghost commented:
> What if the wires are cut?

For other machines if wires are cut does that prevent the ability for items to be pulsed by attachments?
On the scanner it doesn't allow the device to be opened if it's locked so that wouldn't be a good way of sabotaging it.

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

ghost commented:
Changes made - Needs to be tested.
Condensed some of the code a bit better. Removed the icon state check - Will re-add if anyone wants to make a sprite for it.

SheldonCurtiss committed:
Changes to be committed:
	modified:   code/game/machinery/scan_gate.dm

imw-Oculus commented:
Wonderful. A follow-up when this is merged would be to create more variables that the Scanner Gate can detected. Thanks!
